The matrix stack and graphics acceleration.

robotman2412 opened the pull request:
Adds the 2D matrix stack to the window object.
Creates a global 2D matrix stack in case of no specified window.
Adds more options to the framebuffer config to configure the matrix stack.

Current status: Untested, but matrix multiplication functions.

robotman2412 committed:
Start work on matrix stack.

tjclement commented on firmware/components/driver_framebuffer/Kconfig:
```diff
@@ -1,4 +1,4 @@
-menu "Driver: framebuffer support"
+menu "Driver: framebuffer & matrices support"
```
These matrices are not separate from the framebuffer driver, are they? It is my understanding that they are used to transform render layers ("windows"). I see them more as animation properties on windows than as a separate functionality.

tjclement replied:
These matrices are not separate from the framebuffer driver, are they? It is my understanding that they are used to transform render layers ("windows"). I see them more as animation properties on windows than as a separate functionality.

tjclement commented on firmware/components/driver_framebuffer/Kconfig:
```diff
@@ -8,5 +8,17 @@ menu "Driver: framebuffer support"
+		default n
+	config MATRIX_STACK_INITIAL_SIZE
+		depends on DRIVER_FRAMEBUFFER_ENABLE
+		int "Initial size allocated to the matrix stack"
```
What is a matrix stack, and what does this size do?

tjclement commented on firmware/components/driver_framebuffer/driver_framebuffer_window.cpp:
```diff
@@ -83,6 +83,10 @@ Window* driver_framebuffer_window_create(const char* name, uint16_t width, uint1
 		window->buffer = (uint8_t*) heap_caps_malloc(((width*height*PIXEL_SIZE)/8)+1, MALLOC_CAP_8BIT);
 	#endif
 	
+	/* Initialise matrix stack */
```
I try to comment _why_ I do something rather than _what_ I do, because the what is usually apparent and the why is unclear to others

renzenicolai replied:
I agree, useless comment, but so are most of mine so meh, leave it if you want to :-)

tjclement commented on firmware/components/driver_framebuffer/Kconfig:
```diff
@@ -8,5 +8,17 @@ menu "Driver: framebuffer support"
         bool "Swap red and blue"
-        default n
+		default n
+	config MATRIX_STACK_INITIAL_SIZE
```
What is a matrix stack, and what does this size do?

tjclement commented on firmware/components/driver_framebuffer/include/driver_framebuffer_compositor.h:
```diff
@@ -31,8 +32,13 @@ typedef struct Window_t {
 	/* Buffer */
 	uint8_t* buffer;
+
+	/* RobotMan2412's matrix stack */
```
Would be nice to explain here what the matrices will be used for

tjclement reviewed (commented):
Thanks for the update on your work. I think it's cool to have linear transformations in our graphics layer. I hope that we will be able to do animations with it in the future.

I'm not yet sure whether this exact implementation you're proposing is the way to go, but very curious to see where you're going.

robotman2412 commented:
To answer many of your questions:

The matrix stack is indeed not a separate system, but i added it to the name to make it easier to find.

Matrices are used to manipulate the canvas, look at it like warping the cloth you draw on.

The matrix stack functions to be able to recover a transformation matrix you may want to use later on, and functions like any other stack.
A stack is last-on-first-off, which makes it so unique and usefull in many fields, in this case you can say: scale the canvas and move it, then push, apply some other arbitrary transformation, then pop and the other arbitrary transformation will not affect it anymore.
This push/pop effect can be nested by the nature of a stack.

To the stack size:
It is bad to let a stack expand itself infinitely, because errored code will just keep going and make it more difficult to find out what happened.
Therefor, i made an intentional limit which can be changed.

For now, the code is far from done.
I will add more comments here as you asked.

robotman2412 committed:
Basic functionality implemented.
Lines will now be appropriately transformed.
The matrix stack now has a fixed maximum size for simplicity.

renzenicolai commented:
Cool stuff, let's see where this goes.

Small initial hints: please change only the minimum needed, please don't change the driver name and please don't put your name in stuff, you're contributing to a work for badge.team, that it's your work will be aparent from the git commit history.

robotman2412 committed:
Triangle drawing functions, but filled rectangles show a seam.
Only circle drawing remains unchanged and unaffected by transformations.

robotman2412 committed:
Revert some namings.

robotman2412 committed:
Revert "Triangle drawing functions, but filled rectangles show a seam. Only circle drawing remains unchanged and unaffected by transformations."

This reverts commit 9fe51631, which was applied to the wrong branch.

robotman2412 committed:
Reverted some namings.
Triangle drawing now functions, however rotated rectangles may show a seam.
Only circles remain unchanged and unaffected by transformations.

robotman2412 committed:
What is going on?!?!?
Never use revert!
It's evil!!!

robotman2412 committed:
This is really coming together now, example gist:
https://gist.github.com/robotman2412/337720ad9866055448bbb7761f507024
Also, i started a bit of work on slowly working towards 3D.

robotman2412 committed:
Start work on improving circles.
Old circle is still present if fill is set to false.

robotman2412 committed:
Circles and triangles are now much improved, see link.
Old circle is still present if fill is set to false.
Demo gist: https://gist.github.com/robotman2412/337720ad9866055448bbb7761f507024

robotman2412 commented:
I feel that this is _nearly_ ready to be merged.
The 2D functions are very nearly done, and when fully done i advise to merge this as the features afterwards will take much longer to mature.

robotman2412 commented:
Demonstration of the graphics can be found here:
[gist: boats.py](https://gist.github.com/robotman2412/337720ad9866055448bbb7761f507024)

robotman2412 committed:
Fix circles in relation to transformations.
Disable debug stuffs.

renzenicolai commented:
Very, very nice work, however I have some points that I would like to see addressed before I can allow this to be merged:
 - Floating point operations in the basic (2d) drawing functions shouldn't be there
 - You have multiple comments referring to your name like "This file is NOT part of an external library but is rather created and maintained by RobotMan2412 and the badge team." you can probably be assured that no one will touch these files. You can add your name to the top of each file you create if you really want your name to be in there, but please don't "claim" parts of the project: it's a group thing, other people can work on your code, just like you did with "my" framebuffer driver :P
 - Please have all the new functions in a separate file as to not clutter the basic drawing routines file with the new stuff
 - Please make the new functions optional with a menuconfig flag
 - Maybe add different parts of your big new addition as separate PRs, I promise you that we will look at them quickly once we start the merging process, it's just that this PR is very, very big.
 - Please don't add commented out code sections, no matter how "usefull" a commented out line might be to you, if it's commented out then it doesn't work. Comments are for... comments :-)
 - If you see commented out code lines in files that you edit feel free to remove them too as I might have slipped up in the past, this whole thing was thrown together in a hurry

robotman2412 commented:
To respond to your questions:

Floating point operations shouldn't be there for simple 2D?
I don't understand this point.
The matrix stack uses floating point numbers, and allowing floats to be input simply removes some clutter and allows programmers more flexibility.
If you think this is really important, please elaborate.

That's right.
I did forget to remove my name from a few comments.
I did that now.

I was thinking about seperating maybe things like shaders into their own files.
It seems you agree, so that can be done.
This does imply that you want me to make a new file with all new functions, and keep the old file untouched.
I can do this, tough i thought the drawing functions fit quite nicely where they are now.

Making new functions optional seperately?
I think this is more trouble than it's worth:
<ul>
<li>The new circle depends on triangles and the matrix stack.</li>
<li>The new quad depends on triangles.</li>
<li>The new text depends on textured triangles.</li>
<li>Textured triangles depend on textures and shaders.</li>
</ul>
I understand your point, but i strongly advise against seperating these features with on/off.
<br><br>
I do agree that this pull request is quite big.
Then again, the amount of things to change to get fancy new graphics is also big.
I don't blame any of you for it, but your drawing functions were very simple, possibly because of not having a graphics guy.
I noticed very early on that some features which many other libraries had were simply missing.
Which is why i add things like the triangle and quad, they're easy enough to implement and make your life so much easier.
I could theoretically work this into seperate PRs, in chronlogical order, altough my highest priority right now is finishing all the 2D things i want to do.

There is indeed a large commented out block, from when i was working on lines.
I simply didn't remove it yet.
I'll also do that.

If you want me to, i can remove other code comments.

robotman2412 committed:
Text now conforms to translations, getPixel and setPixel can optionally ignore the matrix stack, clearMatrix can now optionally only clear the current matrix.

robotman2412 commented:
I'm satisfied with the current version.
There is some features which i would like to add later, but without further notice this is v1.0.

robotman2412 committed:
Configurable thingies for testing, separate files a bit.

robotman2412 commented:
no wait i broke it

robotman2412 commented:
False alarm, i cleaned build and it worked.

renzenicolai commented on firmware/components/driver_framebuffer/Kconfig:
```diff
@@ -8,5 +8,29 @@ menu "Driver: framebuffer support"
+		default n
+	config MATRIX_STACK_SIZE
+		depends on DRIVER_FRAMEBUFFER_ENABLE
+		int "Size allocated to the matrix stack by default"
```
Size in what? bytes? kilobytes?

renzenicolai commented on firmware/components/driver_framebuffer/driver_framebuffer.cpp:
```diff
@@ -118,6 +120,24 @@ esp_err_t driver_framebuffer_init()
 		driver_hub75_switch_buffer(framebuffer); //Needed to make the legacy compositor work.
 	#endif
 	
+	//Initialise the globl matrix stack.
```
Please remove these comments, might be fun to explain this in the docs, but the code is not the place for this...

renzenicolai commented on firmware/components/driver_framebuffer/driver_framebuffer.cpp:
```diff
@@ -118,6 +120,24 @@ esp_err_t driver_framebuffer_init()
+	//      head
+	//With this hierarchical view, it is easier to see what i mean by this.
+	matrix_stack_2d_init(&stack_2d_global);
+	//later on, i may want to add a 3D matrix stack and a depth buffer, required elements for 3D graphics
```
It's good that you have plans, but again: not here

renzenicolai commented on firmware/components/driver_framebuffer/driver_framebuffer_window.cpp:
```diff
@@ -93,6 +97,8 @@ Window* driver_framebuffer_window_create(const char* name, uint16_t width, uint1
 void driver_framebuffer_window_remove(Window* window)
 {
 	if (window->buffer) free(window->buffer);
+	free(window->stack_2d->matrices);
```
Did you ifdef these?

renzenicolai commented on firmware/components/micropython/esp32/modframebuffer.c:
```diff
@@ -735,10 +1302,10 @@ static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN( framebuffer_window_list_obj,
 
-static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN( framebuffer_get_pixel_obj,            2, 3, framebuffer_get_pixel);
+static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN( framebuffer_get_pixel_obj,            2, 4, framebuffer_get_pixel);
 /* Get the color of a pixel in the framebuffer or in a window. Arguments: window (optional), x, y */
```
Please update the description too

renzenicolai commented on firmware/components/micropython/esp32/modframebuffer.c:
```diff
@@ -735,10 +1302,10 @@ static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN( framebuffer_window_list_obj,
 /* Get the color of a pixel in the framebuffer or in a window. Arguments: window (optional), x, y */
 
-static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN( framebuffer_draw_pixel_obj,           3, 4, framebuffer_draw_pixel);
+static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN( framebuffer_draw_pixel_obj,           3, 5, framebuffer_draw_pixel);
```
Please update the description too

renzenicolai commented on firmware/components/driver_framebuffer/driver_framebuffer.cpp:
```diff
@@ -118,6 +120,24 @@ esp_err_t driver_framebuffer_init()
+	//    maybe we want to animate a figure here
+	//      head
+	//With this hierarchical view, it is easier to see what i mean by this.
+	matrix_stack_2d_init(&stack_2d_global);
```
I sense a missing ifdef? Please make it possible to completely disable the matrix functionality.

renzenicolai reviewed (changes requested):
Close, you're almost there!

I do have some small points though, please check the comments I added.

robotman2412 committed:
Tweaks before merge.
